pkg/filesystem/pool/block_device_backed_file_pool.go

```diff
@@ -129,36 +134,43 @@ func (f *blockDeviceBackedFile) GetNextRegionOffset(off int64, regionType filesy
 	case filesystem.Data:
 		if sectorIndex >= len(f.sectors) {
 			// Inside the hole at the end of the file.
-			return 0, io.EOF
+			return f.holeSource.GetNextRegionOffset(off, regionType)
 		}
 		if f.sectors[sectorIndex] != 0 {
 			// Already inside a sector containing data.
 			return off, nil
 		}
 		// Find the next sector containing data.
-		for {
+		sectorIndex++
+		for f.sectors[sectorIndex] == 0 {
 			sectorIndex++
-			if f.sectors[sectorIndex] != 0 {
-				return int64(sectorIndex) * sectorSizeBytes, nil
-			}
 		}
-	case filesystem.Hole:
-		if sectorIndex >= len(f.sectors) || f.sectors[sectorIndex] == 0 {
-			// Already inside a hole.
-			return off, nil
+		o := int64(sectorIndex) * sectorSizeBytes
```

Comment:
Maybe rename `o` to `fileNextRegionOffset and `u` to `holeSourceNextRegionOffset`?